imap/jmap_ical.c

```diff
@@ -241,13 +241,22 @@ static const char *sha1key(const char *val, char *keybuf)
 
 static char *normalized_uri(const char *uri)
 {
-    const char *col = strchr(uri, ':');
-    if (!col) return xstrdupnull(uri);
+    if (!uri) return NULL;
 
     struct buf buf = BUF_INITIALIZER;
-    buf_setmap(&buf, uri, col - uri);
-    buf_lcase(&buf);
-    buf_appendcstr(&buf, col);
+    const char *col = strchr(uri, ':');
+    if (col) {
+        buf_setmap(&buf, uri, col - uri);
+        buf_lcase(&buf);
+        buf_appendcstr(&buf, col);
+    }
+    else buf_setcstr(&buf, uri);
+
+    buf_trim(&buf);
+    if (!buf_len(&buf)) {
+        buf_free(&buf);
+        return NULL;
+    }
     return buf_release(&buf);
```

Comment:
if we wanted to do the "don't allocate memory if using a short struct buf" magic, we'd need to make APIs like this write into a buf rather than returning a malloced `char *`.  That does mean you'd need to use a flag value for "error" though.

Reply:
yes, changing existing char* return types will require lots of code change. It will be somewhat easy to implement the internal optimization in struct buf. But to really make use of it, we'll have to update existing code. An initial profiling run could tell us where to start looking for the biggest gain.